lighthouse-core/report/html/renderer/snippet-list-renderer.js

```diff
@@ -0,0 +1,307 @@
+    }
+
+    return header;
+  }
+
+  /**
+   * Renders a line of DOM content (code, message, or empty line)
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LineDetails} lineDetails
+   * @return {Element}
+   */
+  static renderSnippetLine(dom, tmpl, {content, number, truncated, contentType, visibility}) {
+    const clonedTemplate = dom.cloneTemplate('#tmpl-lh-snippet__line', tmpl);
+    const codeLine = dom.find('.lh-snippet__line', clonedTemplate);
+    const {classList} = codeLine;
+
+    const typeClasses = {
+      [LineContentType.CODE_NORMAL]: ['lh-snippet__line--content'],
+      [LineContentType.CODE_HIGHLIGHTED]: [
+        'lh-snippet__line--content',
+        'lh-snippet__line--highlighted',
+      ],
+      [LineContentType.EMPTY]: ['lh-snippet__line--empty'],
+      [LineContentType.MESSAGE]: ['lh-snippet__line--message'],
+    }[contentType];
+    typeClasses.forEach(typeClass => classList.add(typeClass));
+
+    if (visibility === LineVisibility.WHEN_COLLAPSED) {
+      classList.add('lh-snippet__show-if-collapsed');
+    }
+    if (visibility === LineVisibility.WHEN_EXPANDED) {
+      classList.add('lh-snippet__show-if-expanded');
+    }
+
+    const lineContent = content + (truncated ? '…' : '');
+    const lineContentEl = dom.find('.lh-snippet__line code', codeLine);
+    if (contentType === LineContentType.MESSAGE) {
+      lineContentEl.appendChild(dom.convertMarkdownLinkSnippets(lineContent));
+    } else {
+      lineContentEl.textContent = lineContent;
+    }
+
+    dom.find('.lh-snippet__line-number', codeLine).textContent = number.toString();
+
+    return codeLine;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererSnippetHighlight} highlight
+   * @return {Element}
+   */
+  static renderHighlightMessage(dom, tmpl, highlight) {
+    return SnippetListRenderer.renderSnippetLine(dom, tmpl, {
+      number: ' ',
+      content: highlight.message,
+      contentType: LineContentType.MESSAGE,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LineVisibility} visibility
+   * @return {Element}
+   */
+  static renderOmittedLinesPlaceholder(dom, tmpl, visibility) {
+    return SnippetListRenderer.renderSnippetLine(dom, tmpl, {
+      number: '…',
+      content: '',
+      visibility,
+      contentType: LineContentType.EMPTY,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererSnippet} details
+   * @return {DocumentFragment}
+   */
+  static renderSnippetContent(dom, tmpl, details) {
+    const {highlights} = details;
+
+    const template = dom.cloneTemplate('#tmpl-lh-snippet__content', tmpl);
+    const snippet = dom.find('.lh-snippet__snippet-inner', template);
+
+    const nonLineSpecificHighlights = highlights.filter(h => typeof h.lineNumber !== 'number');
```

Comment:
Doesn't seem like moving the highlights onto lines is a great improvement:
- `highlights` are what we get from the audit, so we'd have to add them to the lines
- to check if we have highlights we now need to iterate through all lines
- Assuming we still want to allow multiple highlight messages per line and might want to add properties like `type: "error"` or `type: "warning"` in the future we'd need a nested type like `lines[0].highlights[0].text`, rather than just a `higlightMessage` string property

What do you think about splitting highlights into `lineHighlights` and `generalHighlights`?

Or maybe `lineMessages` and `generalMessages` would be better?

Reply:
> Doesn't seem like moving the highlights onto lines is a great improvement

yeah, all your points are good ones

> What do you think about splitting highlights into `lineHighlights` and `generalHighlights`?

sounds good to me, if that seems like a win (I'm fine with either naming scheme)